toffy/qc_comp.py

```diff
@@ -246,209 +248,163 @@ def compute_99_9_intensity(image_data):
-                                                    fovs=batch_names,
-                                                    dtype=dtype)
-
-        # get the channel names directly from image_data if not specified
-        if channels is None:
-            channels = image_data.channels.values
-
-        # verify the channel names (important if the user explicitly specifies channels)
-        misc_utils.verify_in_list(
-            provided_chans=channels,
-            image_chans=image_data.channels.values
-        )
-
-        # compute the QC metrics of this batch
-        qc_data_batch = compute_qc_metrics_batch(
-            image_data, batch_names, channels, gaussian_blur, blur_factor
-        )
-
-        # append the batch QC metric data to the full processed data
-        df_nonzero_mean = pd.concat(
-            [df_nonzero_mean, qc_data_batch['nonzero_mean_batch']]
-        )
-        df_total_intensity = pd.concat(
-            [df_total_intensity, qc_data_batch['total_intensity_batch']]
-        )
-        df_99_9_intensity = pd.concat(
-            [df_99_9_intensity, qc_data_batch['99_9_intensity_batch']]
-        )
-
-        # update number of fovs processed
-        fovs_processed += batch_size
-
-        # print fovs processed update
-        print("Number of fovs processed: %d" % fovs_processed)
-
-    # reset the indices to make indexing consistent
-    df_nonzero_mean = df_nonzero_mean.reset_index(drop=True)
-    df_total_intensity = df_total_intensity.reset_index(drop=True)
-    df_99_9_intensity = df_99_9_intensity.reset_index(drop=True)
-
-    # create a dictionary mapping the metric name to its respective DataFrame
-    qc_data = {
-        'nonzero_mean': df_nonzero_mean,
-        'total_intensity': df_total_intensity,
-        '99_9_intensity': df_99_9_intensity
-    }
-
-    return qc_data
+    # path validation check
+    if not os.path.exists(bin_file_path):
+        raise FileNotFoundError('bin_file_path %s does not exist' % bin_file_path)
+
+    # iterate over each metric substr
+    metric_substrs = [
+        'nonzero_mean_stats.csv', 'total_intensity_stats.csv', 'percentile_99_9_stats.csv'
+    ]
+    for ms in metric_substrs:
+        # define an aggregated metric DataFrame
+        metric_df = pd.DataFrame()
+
+        # list all the files corresponding to this metric
+        metric_files = io_utils.list_files(bin_file_path, substrs=ms)
+
+        # this prevents an already-existing combined .csv file from sneaking in
+        metric_files = [
+            mf_match[0] + '_%s' % ms for mf in metric_files if
+            len(mf_match := re.findall(r'fov-\d+-scan-\d+', mf)) == 1
+        ]
+
+        # sort the files to ensure consistency
+        # NOTE: all FOVs are named fov-m-scan-n, 'm' determines run acquisition order
+        # TODO: need an additional sort by scan?
+        metric_files = sorted(metric_files, key=lambda x: re.findall(r'\d+', x)[0])
```

Comment:
We should pull this out into a helper function that works from `m` from 1 to 101 and `n` from 1 to 99. Other functions will need to sort FOVs by order

Reply:
In other words, toss out any bin file FOV name that has a fov number outside `range(1, 102) and scan number outside `range(1, 100)` when sorting?

Reply:
It shouldn't out numbers outside that range, but it should definitely work for numbers inside that ranges. I don't expect we'll ever have four digits worth of bin files or three digits worth of scans